[SPARK-27468][core] Track correct storage level of RDDs and partitions.

vanzin opened the pull request:
Previously, the RDD level would change depending on the status reported
by executors for the block they were storing, and individual blocks would
reflect that. That is wrong because different blocks may be stored differently
in different executors.

So now the RDD tracks the user-provided storage level, while the individual
partitions reflect the current storage level of that particular block,
including the current number of replicas.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/15613/
Test PASSed.

vanzin commented:
Ping? Anyone? Bueller?

vanzin commented:
Let's try... @squito @attilapiros

SparkQA commented:
**[Test build #111456 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111456/testReport)** for PR 25779 at commit [`effcbb5`](https://github.com/apache/spark/commit/effcbb5c2be4d980e8e818571e58acad199ee7b3).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16490/
Test PASSed.

SparkQA commented:
**[Test build #111456 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111456/testReport)** for PR 25779 at commit [`effcbb5`](https://github.com/apache/spark/commit/effcbb5c2be4d980e8e818571e58acad199ee7b3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111456/
Test PASSed.

SparkQA commented:
**[Test build #111489 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111489/testReport)** for PR 25779 at commit [`d2c5ed4`](https://github.com/apache/spark/commit/d2c5ed4c2b93b6852130909bb453a7996b90472d).

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16517/
Test PASSed.

SparkQA commented:
**[Test build #111489 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111489/testReport)** for PR 25779 at commit [`d2c5ed4`](https://github.com/apache/spark/commit/d2c5ed4c2b93b6852130909bb453a7996b90472d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class _GaussianMixtureParams(HasMaxIter, HasFeaturesCol, HasSeed, HasPredictionCol,`
  * `class GaussianMixtureModel(JavaModel, _GaussianMixtureParams, JavaMLWritable, JavaMLReadable,`
  * `class GaussianMixture(JavaEstimator, _GaussianMixtureParams, JavaMLWritable, JavaMLReadable):`
  * `class _KMeansParams(HasMaxIter, HasFeaturesCol, HasSeed, HasPredictionCol, HasTol,`
  * `class KMeansModel(JavaModel, _KMeansParams, GeneralJavaMLWritable, JavaMLReadable,`
  * `class KMeans(JavaEstimator, _KMeansParams, JavaMLWritable, JavaMLReadable):`
  * `class _BisectingKMeansParams(HasMaxIter, HasFeaturesCol, HasSeed, HasPredictionCol,`
  * `class BisectingKMeansModel(JavaModel, _BisectingKMeansParams, JavaMLWritable, JavaMLReadable,`
  * `class BisectingKMeans(JavaEstimator, _BisectingKMeansParams, JavaMLWritable, JavaMLReadable):`
  * `class _LDAParams(HasMaxIter, HasFeaturesCol, HasSeed, HasCheckpointInterval):`
  * `class LDAModel(JavaModel, _LDAParams):`
  * `class LDA(JavaEstimator, _LDAParams, JavaMLReadable, JavaMLWritable):`
  * `class _PowerIterationClusteringParams(HasMaxIter, HasWeightCol):`
  * `class PowerIterationClustering(_PowerIterationClusteringParams, JavaParams, JavaMLReadable,`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111489/
Test PASSed.

vanzin commented on core/src/test/scala/org/apache/spark/status/LiveEntitySuite.scala:
```diff
@@ -51,6 +52,61 @@ class LiveEntitySuite extends SparkFunSuite {
+    assert(exec.usedOnHeap === 1)
+    assert(exec.usedOffHeap === 0)
+
+    val old1 = exec.addBlock(RDDBlockId(1, 1), 3L, 3L, false)
```
IIRC, if a block is evicted from memory to disk, for example; and later if it's loaded back into memory.

squito replied:
ah good point.  worth a comment?  that case usually slips my mind, anyway

vanzin commented on core/src/test/scala/org/apache/spark/status/AppStatusListenerSuite.scala:
```diff
@@ -763,6 +769,7 @@ class AppStatusListenerSuite extends SparkFunSuite with BeforeAndAfter {
       assert(part.memoryUsed === rdd1b1.memSize * 2)
```
Sure, I can add an assert.

The list is not ordered for performance reasons. See `RDDPartitionSeq` in LiveEntyty.scala.

vanzin replied:
(actually the assert / change is not needed since a few lines above there's an assertion that the list has a single element.)

squito replied:
ok fair enough.

I did realize why RDDPartitionSeq is the way it is after a closer look, I was just mentioning that its not so obvious, or commented in the places I expected (the api just exposes a `Seq[RDDPartitionInfo]`)

vanzin commented on core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:
```diff
@@ -235,13 +230,14 @@ private[spark] class AppStatusListener(
             }
           }
+        }
         update(rdd, now)
```
Hmm, not sure where you're seeing the conditional update? This line is executed for all live RDDs in the old code.

attilapiros replied:
I guess the `update` in this line was unnecessary? 

https://github.com/apache/spark/blob/4c8f114783241e780c0ad227bfb3191657637df7/core/src/main/scala/org/apache/spark/status/AppStatusListener.scala#L219-L221

I think this is the `update` which caused the confusion for @squito too.

squito replied:
right, it was that part, combined with misreading this part of the old code.  Though it seems like it could be a conditional update, but no change from before.

SparkQA commented:
**[Test build #111504 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111504/testReport)** for PR 25779 at commit [`f6a539a`](https://github.com/apache/spark/commit/f6a539aa1726753b2c3e239a0bb071975e170eb3).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16526/
Test PASSed.

SparkQA commented:
**[Test build #111504 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111504/testReport)** for PR 25779 at commit [`f6a539a`](https://github.com/apache/spark/commit/f6a539aa1726753b2c3e239a0bb071975e170eb3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111504/
Test FAILed.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #111611 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111611/testReport)** for PR 25779 at commit [`f6a539a`](https://github.com/apache/spark/commit/f6a539aa1726753b2c3e239a0bb071975e170eb3).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16610/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111611/
Test FAILed.

attilapiros reviewed (commented):
Now it was just a quick going through on the changes and I would like to read it through more carefully tomorrow (CET time).

vanzin commented on core/src/main/scala/org/apache/spark/status/LiveEntity.scala:
```diff
@@ -269,13 +278,127 @@ private class LiveExecutor(val executorId: String, _addTime: Long) extends LiveE
+   * This is a var so that we can "clear" it and restart with a new map when it becomes empty,
+   * to save some memory.
+   */
+  private var rddBlockUsage = new OpenHashMap[Int, RDDBlockTracker]()
```
No. You also missed the comment on the very previous line.

attilapiros replied:
Yes I missed that one. It is strange the map is not clearable in this case this is very fine this way.

squito reviewed (approved):
lgtm

attilapiros commented on core/src/main/scala/org/apache/spark/status/LiveEntity.scala:
```diff
@@ -524,30 +648,35 @@ private class LiveRDD(val info: RDDInfo) extends LiveEntity {
-  private val partitions = new HashMap[String, LiveRDDPartition]()
+  private val partitions = new OpenHashMap[Int, LiveRDDPartition]()
   private val partitionSeq = new RDDPartitionSeq()
+  private var partitionCount = 0
```
Why a separate `partitionCount` is needed? The `partitionSeq` has a length method which follows the add/remove partition.

attilapiros commented on core/src/test/scala/org/apache/spark/status/AppStatusListenerSuite.scala:
```diff
@@ -800,9 +807,27 @@ class AppStatusListenerSuite extends SparkFunSuite with BeforeAndAfter {
     // Remove block 1 from bm 1.
     listener.onBlockUpdated(SparkListenerBlockUpdated(
-      BlockUpdatedInfo(bm1, rdd1b1.blockId, StorageLevel.NONE, rdd1b1.memSize, rdd1b1.diskSize)))
+      BlockUpdatedInfo(bm1, rdd1b1.blockId, StorageLevel.NONE, 0L, 0L)))
```
This was I assume a mistake in the test as the delta was reported here and not the currently used mem and disk size. This is why this line is changed.

SparkQA commented:
**[Test build #111623 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111623/testReport)** for PR 25779 at commit [`055da38`](https://github.com/apache/spark/commit/055da3845bf625cb66a21c9b193f69ca247f5e17).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16621/
Test PASSed.

SparkQA commented:
**[Test build #111623 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111623/testReport)** for PR 25779 at commit [`055da38`](https://github.com/apache/spark/commit/055da3845bf625cb66a21c9b193f69ca247f5e17).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111623/
Test FAILed.

HeartSaVioR commented:
We might need to inform about the change of LiveEntity to the patch of #25943 , otherwise we might forget to include it in snapshot.

So according to the javadoc, `BlockManager.reportBlockStatus` is intended to current storage status of a block, which is clear, except the case of `BlockManager.dropFromMemory`. It also receives droppedMemorySize and let master deal with information of both sizes (memory/disk) which is really confusing and doesn't match of the explanation of "block managers report the current mem and disk used by the block".

Could we remove the exceptional case as well, or there's technical reason to leave it as it is?

vanzin commented:
> It also receives droppedMemorySize and let master deal with information of both sizes (memory/disk) which is really confusing

Yeah, I'll have to take a second look at that code, run some more experiments. That's a really awkward API... let me see if there's a way to end up with the correct info without having to change the block manager code.

vanzin commented:
I'll take a closer look tomorrow, but a quick experiment tells me that:

- if the block is dropped from memory while adding another block, the update contains a storage level with "useMemory = false", and the correct memory size that was dropped
- if the block is dropped via `BlockManager.removeBlockInternal`, which is triggered by unpersisting an RDD, then the block update sent to the master does *not* contain the previous size, neither for disk nor memory.

The thing with the second case is that the code path that triggers it doesn't send an update to the master. I'm actually working on a feature where that would happen, though.

But given that, I'll probably undo some of the changes here. Mainly relying on the fact that the block update contains the delta for the memory, and that disk blocks are never dropped outside of unpersist events.

For my WIP changes (for which I'll eventually post a PR) I'll look at reporting the correct sizes when dropping blocks in the second case above.

Commit:
[SPARK-27468][core] Track correct storage level of RDDs and partitions.

Previously, the RDD level would change depending on the status reported
by executors for the block they were storing, and individual blocks would
reflect that. That is wrong because different blocks may be stored differently
in different executors.

So now the RDD tracks the user-provided storage level, while the individual
partitions reflect the current storage level of that particular block,
including the current number of replicas.

vanzin commented:
(BTW it could be that the problem in the second bullet above is fixed by #25973. I'll look more carefully tomorrow.)

HeartSaVioR commented:
> if the block is dropped via BlockManager.removeBlockInternal, which is triggered by unpersisting an RDD, then the block update sent to the master does not contain the previous size, neither for disk nor memory.

Right, #25973 is for dealing with this, though the origin target of #25973 was broadcast.

So if we would want to allow two different semantics of "size" on UpdateBlockInfo (either "addition" and "removal" depending on storage level), I guess #25973 is correct. If we feel awkward about current logic and only want to report "current" status of block, we may have to make non trivial change and #25973 will be no longer correct then.

vanzin commented:
Ok, running some tests I found a couple of extra bugs. I'll file a new bug with the details, and then update this one with the fix for SPARK-27468 specifically.

After my investigations, #25973 seems correct, but it does not fix the new bugs that I found in my testing.

vanzin commented:
Filed SPARK-29319. I'll update this with the fixes that aren't related to that bug.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16648/
Test PASSed.

SparkQA commented:
**[Test build #111652 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111652/testReport)** for PR 25779 at commit [`0848fbf`](https://github.com/apache/spark/commit/0848fbfcac7bb100962c20bcd83bd7fe1a6e0a46).

SparkQA commented:
**[Test build #111652 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111652/testReport)** for PR 25779 at commit [`0848fbf`](https://github.com/apache/spark/commit/0848fbfcac7bb100962c20bcd83bd7fe1a6e0a46).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111652/
Test PASSed.

vanzin commented:
Ping

squito commented on core/src/main/scala/org/apache/spark/status/LiveEntity.scala:
```diff
@@ -520,27 +521,28 @@ private class LiveRDDDistribution(exec: LiveExecutor) {
+ * Tracker for data related to a persisted RDD. Note the storage level is kept immutable here,
+ * following the current behavior of `RDD.persist()`, even though it is mutable in the `RDDInfo`
+ * structure.
+ */
```
A stage may have been submitted before the RDD was persisted at all, then another stage submitted after the RDD is persisted, so its not actually immutable.  You wouldn't properly capture that here.

(probably not the optimal thing for the user to do, but I've seen weirder things ...)

vanzin replied:
Actually that works fine, because the listener does not track RDDs that are not persisted; so there wouldn't be a live RDD for the first stage in your example; it would be created when the second stage is submitted, and at that point the storage level cannot be changed further.

I'll update the comment (maybe even add a unit test).

squito replied:
makes sense, thanks for adding the test too.

squito commented on core/src/main/scala/org/apache/spark/status/LiveEntity.scala:
```diff
@@ -458,7 +458,7 @@ private class LiveStage extends LiveEntity {
 }
 
-private class LiveRDDPartition(val blockName: String) {
+private class LiveRDDPartition(val blockName: String, rddLevel: StorageLevel) {
```
It would help me a bit if this was called 'requestedStorageLevel' and in RDDPartitionInfo it was called 'effectiveStorageLevel'.  I guess its not worth changing RDDPartitionInfo, but maybe just a comment along those lines.

vanzin replied:
Sure.

Commit:
Feedback.

SparkQA commented:
**[Test build #111792 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111792/testReport)** for PR 25779 at commit [`b95baec`](https://github.com/apache/spark/commit/b95baec2297254707fecf6a819ecd86b088a68d9).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16774/
Test PASSed.

SparkQA commented:
**[Test build #111792 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111792/testReport)** for PR 25779 at commit [`b95baec`](https://github.com/apache/spark/commit/b95baec2297254707fecf6a819ecd86b088a68d9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111792/
Test PASSed.

squito reviewed (approved):
lgtm

squito commented:
merged to master

dongjoon-hyun commented:
Hi, @vanzin and @squito and @zsxwing .

SPARK-27468 is reported by @zsxwing at 2.4.1. Can we have this bug fix in `branch-2.4` because Apache Spark 2.4 is LTS?

Also, cc @srowen
